TM1py/Services/CellService.py

```diff
@@ -368,6 +370,57 @@ def write_dataframe(self, cube_name: str, data: 'pd.DataFrame', dimensions: Iter
                           use_changeset=use_changeset,
                           **kwargs)
 
+    @require_pandas
+    @manage_transaction_log
+    def write_dataframe_async(self, cube_name: str, data: 'pd.DataFrame',
+                              slice_size_of_dataframe: int, max_workers: int,
+                              increment: bool = False):
+        """ Write DataFrame into a cube using unbound TI processes in a multi-threading way. Requires admin permissions.
+        For a DataFrame with > 1,000,000 rows, this function will at least save half of runtime compared with `write_dataframe` function.
+        Column order must match dimensions in the target cube with an additional column for the values.
+        Column names are not relevant.
+        :param cube_name:
+        :param data: Pandas Data Frame
+        :param slice_size_of_dataframe: Number of rows for each DataFrame slice, e.g. 10000
+        :param max_workers: Max number of threads, e.g. 14
+        :param increment: increment or update cell values. Defaults to False.
+        :return: the Future’s result or raise exception.
+        """
+        def chunks(data):
+            slice_list = [data.iloc[i:i + slice_size_of_dataframe] for i in range(0, data.shape[0], slice_size_of_dataframe)]
+            return slice_list
+
+        def write(self, chunk):
+            try:
+                return self.write_dataframe(cube_name=cube_name, data=chunk, increment=increment, use_ti=True)
+            except TM1pyException as e:
+                raise e
+
+        async def write_async(self, data):
+            loop = asyncio.get_event_loop()
+            outcomes = []
+
+            with ThreadPoolExecutor(max_workers) as executor:
+                futures = [loop.run_in_executor(executor, write, self, chunk) for chunk in chunks(data)]
+
+                for future in futures:
+                    outcomes.append(await future)
+
+            return outcomes
+
+        if not isinstance(data, pd.DataFrame):
```

Comment:
This could be moved to the top this method.

Reply:
Yes, agreed.